samplers/jaegerremote/sampler_remote.go

```diff
@@ -0,0 +1,313 @@
+	sampler := &Sampler{
+		samplerOptions: *options,
+		serviceName:    serviceName,
+		doneChan:       make(chan *sync.WaitGroup),
+	}
+	go sampler.pollController()
+	return sampler
+}
+
+func (s *Sampler) ShouldSample(p trace.SamplingParameters) trace.SamplingResult {
+	s.RLock()
+	defer s.RUnlock()
+	return s.sampler.ShouldSample(p)
+}
+
+// Close does a clean shutdown of the sampler, stopping any background
+// go-routines it may have started.
+func (s *Sampler) Close() {
+	if swapped := atomic.CompareAndSwapInt64(&s.closed, 0, 1); !swapped {
+		otel.Handle(fmt.Errorf("repeated attempt to close the sampler is ignored"))
+		return
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	s.doneChan <- &wg
+	wg.Wait()
+}
+
+func (s *Sampler) Description() string {
+	return "JaegerRemoteSampler{}"
+}
+
+func (s *Sampler) pollController() {
+	ticker := time.NewTicker(s.samplingRefreshInterval)
+	defer ticker.Stop()
+	s.pollControllerWithTicker(ticker)
+}
+
+func (s *Sampler) pollControllerWithTicker(ticker *time.Ticker) {
+	for {
+		select {
+		case <-ticker.C:
+			s.UpdateSampler()
+		case wg := <-s.doneChan:
+			wg.Done()
+			return
+		}
+	}
+}
+
+func (s *Sampler) setSampler(sampler trace.Sampler) {
+	s.Lock()
+	defer s.Unlock()
+	s.sampler = sampler
+}
+
+// UpdateSampler forces the sampler to fetch sampling strategy from backend server.
+// This function is called automatically on a timer, but can also be safely called manually, e.g. from tests.
+func (s *Sampler) UpdateSampler() {
+	res, err := s.samplingFetcher.Fetch(s.serviceName)
+	if err != nil {
+		//s.logger.Infof("failed to fetch sampling strategy: %v", err)
+		return
+	}
+	strategy, err := s.samplingParser.Parse(res)
+	if err != nil {
+		//s.logger.Infof("failed to parse sampling strategy response: %v", err)
+		return
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	if err := s.updateSamplerViaUpdaters(strategy); err != nil {
+		//s.logger.Infof("failed to handle sampling strategy response %+v. Got error: %v", res, err)
+		return
+	}
+}
+
+// NB: this function should only be called while holding a Write lock
+func (s *Sampler) updateSamplerViaUpdaters(strategy interface{}) error {
+	for _, updater := range s.updaters {
+		sampler, err := updater.Update(s.sampler, strategy)
+		if err != nil {
+			return err
+		}
+		if sampler != nil {
+			s.sampler = sampler
+			return nil
+		}
+	}
+	return fmt.Errorf("unsupported sampling strategy %+v", strategy)
+}
+
+// -----------------------
+
+// probabilisticSamplerUpdater is used by Sampler to parse sampling configuration.
+type probabilisticSamplerUpdater struct{}
+
+// Update implements Update of samplerUpdater.
+func (u *probabilisticSamplerUpdater) Update(sampler trace.Sampler, strategy interface{}) (trace.Sampler, error) {
+	type response interface {
+		GetProbabilisticSampling() *jaeger_api_v2.ProbabilisticSamplingStrategy
+	}
+	var _ response = new(jaeger_api_v2.SamplingStrategyResponse) // sanity signature check
```

Comment:
Isn't this a compile-time check that can be performed outside the context of a function invocation?

Reply:
Can't be outside because `response` interface is defined in-scope. The compiler will not generate any code for this.